File: bindings/java/src/main/java/com/aster/interceptors/RetryInterceptor.java
package com.aster.interceptors;

import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Retry interceptor providing retry policy hints for client calls.
 *
 * <p>This interceptor does not perform retries itself -- it exposes {@link #shouldRetry} and {@link
 * #backoffSeconds} for the client dispatch loop to consult.
 */
public final class RetryInterceptor implements Interceptor {

  private final Set<StatusCode> retryableCodes;
  private final long initialMs;
  private final long maxMs;
  private final double multiplier;
  private final double jitter;

  /**
   * Creates a retry interceptor with the given backoff parameters.
   *
   * @param retryableCodes status codes eligible for retry (default: UNAVAILABLE only)
   * @param initialMs initial backoff in milliseconds (default 100)
   * @param maxMs maximum backoff in milliseconds (default 5000)
   * @param multiplier exponential backoff multiplier (default 2.0)
   * @param jitter jitter factor (0.0 to 1.0) applied to the delay (default 0.2)
   */
  public RetryInterceptor(
      Set<StatusCode> retryableCodes,
      long initialMs,
      long maxMs,
      double multiplier,
      double jitter) {
    this.retryableCodes = retryableCodes != null ? retryableCodes : Set.of(StatusCode.UNAVAILABLE);
    this.initialMs = initialMs;
    this.maxMs = maxMs;
    this.multiplier = multiplier;
    this.jitter = jitter;
  }

  /** Creates a retry interceptor with default settings. */
  public RetryInterceptor() {
    this(Set.of(StatusCode.UNAVAILABLE), 100, 5000, 2.0, 0.2);
  }

  /**
   * Returns {@code true} if the call should be retried.
   *
   * <p>A call is retryable if it is marked idempotent and the error code is in the set of retryable
   * codes.
   */
  public boolean shouldRetry(CallContext ctx, RpcError error) {
    return ctx.isIdempotent() && retryableCodes.contains(error.code());
  }

  /**
   * Computes the backoff delay in seconds for the given attempt number.
   *
   * <p>Uses exponential backoff with jitter: {@code min(maxMs, initialMs * multiplier^(attempt-1))
   * + jitter}.
   *
   * @param attempt the retry attempt number (1-based)
   * @return delay in seconds before the next retry
   */
  public double backoffSeconds(int attempt) {
    long delayMs =
        Math.min(maxMs, (long) (initialMs * Math.pow(multiplier, Math.max(0, attempt - 1))));
    double jitterMs = delayMs * jitter * ThreadLocalRandom.current().nextDouble();
    return (delayMs + jitterMs) / 1000.0;
  }
}

File: bindings/dotnet/src/Aster/Interceptors/RetryInterceptor.cs
namespace Aster.Interceptors;

/// <summary>
/// Provides retry policy hints for client calls.
/// Does not retry automatically -- exposes ShouldRetry and BackoffSeconds
/// for the call loop to consult.
/// </summary>
public class RetryInterceptor : IInterceptor
{
    /// <summary>Maximum number of attempts (including the first).</summary>
    public int MaxAttempts { get; }

    /// <summary>Status codes considered retryable.</summary>
    public HashSet<StatusCode> RetryableCodes { get; }

    // Backoff parameters
    private readonly int _initialMs;
    private readonly int _maxMs;
    private readonly double _multiplier;
    private readonly double _jitter;

    private static readonly Random Rng = new();

    /// <param name="maxAttempts">Maximum number of attempts (including the first). Defaults to 3.</param>
    /// <param name="retryableCodes">Status codes considered retryable. Defaults to {Unavailable}.</param>
    /// <param name="initialMs">Initial backoff delay in milliseconds. Defaults to 100.</param>
    /// <param name="maxMs">Maximum backoff delay in milliseconds. Defaults to 30000.</param>
    /// <param name="multiplier">Multiplicative factor per attempt. Defaults to 2.0.</param>
    /// <param name="jitter">Random jitter factor (0.0-1.0). Defaults to 0.1.</param>
    public RetryInterceptor(
        int maxAttempts = 3,
        HashSet<StatusCode>? retryableCodes = null,
        int initialMs = 100,
        int maxMs = 30_000,
        double multiplier = 2.0,
        double jitter = 0.1)
    {
        MaxAttempts = maxAttempts;
        RetryableCodes = retryableCodes ?? new HashSet<StatusCode> { StatusCode.Unavailable };
        _initialMs = initialMs;
        _maxMs = maxMs;
        _multiplier = multiplier;
        _jitter = jitter;
    }

    /// <summary>
    /// Whether the given error should be retried based on the call context.
    /// </summary>
    public bool ShouldRetry(CallContext ctx, RpcError error)
    {
        return ctx.Idempotent && RetryableCodes.Contains(error.Code);
    }

    /// <summary>
    /// Compute the backoff delay in seconds for the given attempt number.
    /// </summary>
    public double BackoffSeconds(int attempt)
    {
        int delayMs = Math.Min(
            _maxMs,
            (int)(_initialMs * Math.Pow(_multiplier, Math.Max(0, attempt - 1))));
        double jitterMs = delayMs * _jitter * Rng.NextDouble();
        return (delayMs + jitterMs) / 1000.0;
    }
}
